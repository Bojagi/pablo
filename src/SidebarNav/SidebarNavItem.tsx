import React from 'react';
import styled, { css } from 'styled-components';
import { buttonBaseStyles, ButtonBaseProps } from '../ButtonBase';
import { getComponentStyle } from '../utils/styleHelpers/getComponentStyle';
import { Typography } from '../Typography';

export interface SidebarNavItemProps extends ButtonBaseProps {
  selected?: boolean;
  children: React.ReactNode;
}

const SidebarNavItemWrapper = styled.li.attrs({ size: 'small' })<SidebarNavItemProps>`
  ${buttonBaseStyles}
  display: flex;
  margin: ${getComponentStyle('sidebarNav.item.marginY')} 0;
  ${(props) =>
    props.selected &&
    css`
      background-color: ${getComponentStyle('sidebarNav.item.selected.backgroundColor')};
    `}

  &:hover {
    background-color: ${getComponentStyle('sidebarNav.item.hover.backgroundColor')};
  }

  &:focus {
    box-shadow: 0 0 0 ${getComponentStyle('sidebarNav.item.focus.outlineSize')}
      ${getComponentStyle('sidebarNav.item.focus.outlineColor')};
  }

  &:active {
    background-color: ${getComponentStyle('sidebarNav.item.active.backgroundColor')};
  }
`;

export const SidebarNavItem = ({ children, ...props }: SidebarNavItemProps) => (
  <SidebarNavItemWrapper data-testid="pbl-sidebarnav-item" {...props}>
    <Typography variant="button">{children}</Typography>
  </SidebarNavItemWrapper>
);
