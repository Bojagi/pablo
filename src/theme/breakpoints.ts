export type Breakpoint = 'sm' | 'md' | 'lg' | 'xl';

export const breakpointNames: Breakpoint[] = ['sm', 'md', 'lg', 'xl'];

export type BreakpointsArray = [number, number, number, number];

export type Breakpoints = BreakpointsArray &
  Record<Breakpoint, number> & {
    breakpointNames: Breakpoint[];
  };

export function createBreakpoints(
  array: BreakpointsArray,
  usedBreakpointNames: Breakpoint[] = breakpointNames
): Breakpoints {
  const newBreakpoints = ([...array] as unknown) as Breakpoints;
  usedBreakpointNames.forEach((bp, index) => {
    newBreakpoints[bp] = newBreakpoints[index];
  });
  newBreakpoints.breakpointNames = usedBreakpointNames;

  return newBreakpoints as Breakpoints;
}

export const defaultBreakpoints: BreakpointsArray = [600, 1000, 1200, 1920];

export const breakpoints = createBreakpoints(defaultBreakpoints, breakpointNames);
