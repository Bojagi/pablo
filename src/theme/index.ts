export * from './defaultTheme';
export * from './context';
export * from './PabloThemeProvider';
export * from './useComponentStyle';
