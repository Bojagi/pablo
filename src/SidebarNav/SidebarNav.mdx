---
name: SidebarNav
route: /sidebarnav
menu: Components
---

import { Props, Playground } from 'docz'
import { Box } from '../Box/Box.tsx'
import { SidebarNav } from './SidebarNav.tsx'
import { SidebarNavItem } from './SidebarNavItem.tsx'

# SidebarNav

The `SidebarNav` component is a navigation that supposed to be placed on the right of the main content.

This component is subject to change and be split up in a `SideBar` component that can host
multiple different components and a vertical navigation.

## Example

<Playground>
  <SidebarNav>
    <SidebarNavItem>Hello</SidebarNavItem>
    <SidebarNavItem selected>Goodbye</SidebarNavItem>
    <SidebarNavItem>Ob1-La-Di Ob-La-Da</SidebarNavItem>
  </SidebarNav>
</Playground>

## Import

```js
import { SidebarNav, SidebarNavItem } from '@bojagi/pablo/SidebarNav';
```

## Props

### SidebarNav

<Props of={SidebarNav} />

### SidebarNavItem

<Props of={SidebarNavItem} />
