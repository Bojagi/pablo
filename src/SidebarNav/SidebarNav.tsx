import React, { forwardRef } from 'react';
import styled from 'styled-components';
import { layoutInterpolationFn, LayoutBoxProps } from '../Box';
import { getComponentStyle } from '../styleHelpers';
import { SidebarNavItemProps } from './SidebarNavItem';

export interface SidebarNavProps extends LayoutBoxProps {
  children: React.ReactElement<SidebarNavItemProps> | React.ReactElement<SidebarNavItemProps>[];
}

const SidebarNavBox = styled.ul<LayoutBoxProps>`
  border-left: ${getComponentStyle('sidebarNav.borderLeft')};
  padding-left: ${getComponentStyle('sidebarNav.borderLeftSpacing')};
  ${layoutInterpolationFn};
`;

export const SidebarNav = forwardRef<HTMLUListElement, SidebarNavProps>(
  ({ children, ...props }: SidebarNavProps, ref) => (
    <SidebarNavBox ref={ref} data-testid="pbl-sidebarnav" {...props}>
      {children}
    </SidebarNavBox>
  )
);
