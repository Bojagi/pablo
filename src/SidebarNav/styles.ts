import { css } from 'styled-components';
import { getColor } from '../utils/styleHelpers/getColor';
import { getSpacing } from '../utils/styleHelpers/getSpacing';
import { Style } from '../theme/types';

export interface SidebarNavItemStyles {
  marginY: Style;
  active: {
    backgroundColor: Style;
    outlineSize: Style;
  };
  focus: {
    outlineColor: Style;
    outlineSize: Style;
  };
  selected: {
    backgroundColor: Style;
  };
  hover: {
    backgroundColor: Style;
  };
}

export interface SidebarNavStyles {
  borderLeft: Style;
  borderLeftSpacing: Style;
  item: SidebarNavItemStyles;
}

export const sidebarNavStyles: SidebarNavStyles = {
  borderLeft: css`1px solid ${getColor('borders', 'light')}`,
  borderLeftSpacing: getSpacing(6),
  item: {
    marginY: getSpacing(4),
    focus: {
      outlineColor: getColor('brand', 'lightest'),
      outlineSize: getSpacing(2),
    },
    active: {
      backgroundColor: getColor('brand', 'light'),
      outlineSize: getSpacing(2),
    },
    selected: {
      backgroundColor: getColor('brand', 'lightest'),
    },
    hover: {
      backgroundColor: getColor('brand', 'lightest'),
    },
  },
};
