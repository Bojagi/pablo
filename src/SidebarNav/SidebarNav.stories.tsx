import React from 'react';
import { SidebarNav } from './SidebarNav';
import { SidebarNavItem } from './SidebarNavItem';

export default {
  title: 'SidebarNav',
};

export const UnselectedNavigation = () => (
  <SidebarNav>
    <SidebarNavItem>Hello</SidebarNavItem>
    <SidebarNavItem>Goodbye</SidebarNavItem>
    <SidebarNavItem>Ob-La-Di Ob-La-Da</SidebarNavItem>
  </SidebarNav>
);

export const SelectedNavigation = () => (
  <SidebarNav>
    <SidebarNavItem>Hello</SidebarNavItem>
    <SidebarNavItem active>Goodbye</SidebarNavItem>
    <SidebarNavItem>Ob-La-Di Ob-La-Da</SidebarNavItem>
  </SidebarNav>
);
