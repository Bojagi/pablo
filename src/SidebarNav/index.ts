import './styles';

export * from './SidebarNav';
export * from './SidebarNavItem';
