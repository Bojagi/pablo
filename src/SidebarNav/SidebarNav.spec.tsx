import { render, cleanup } from '@testing-library/react';
import React from 'react';
import { defaultTheme, PabloThemeProvider } from '../theme';
import { SidebarNav } from './SidebarNav';
import { SidebarNavItem } from './SidebarNavItem';

let onSelectMock;

beforeEach(() => {
  onSelectMock = jest.fn();
});

afterEach(cleanup);

test('Render one nav item', () => {
  const { container } = render(
    <PabloThemeProvider>
      <SidebarNav>
        <SidebarNavItem selected>Hello</SidebarNavItem>
      </SidebarNav>
    </PabloThemeProvider>
  );
  expect(container).toMatchSnapshot();
});

test('Render multiple nav items', () => {
  const { container, getAllByTestId } = renderComponent({});
  expect(container).toMatchSnapshot();
  const allSidebarNav = getAllByTestId('pbl-sidebarnav-item');
  expect(allSidebarNav).toHaveLength(3);
  expect(allSidebarNav[0]).toHaveTextContent('Hello');
  expect(allSidebarNav[1]).toHaveTextContent('World');
  expect(allSidebarNav[2]).toHaveTextContent('Bye');
  expect(allSidebarNav[0]).toHaveStyleRule('background-color', undefined);
  expect(allSidebarNav[1]).toHaveStyleRule('background-color', defaultTheme.colors.brand.lightest);
  expect(allSidebarNav[2]).toHaveStyleRule('background-color', undefined);
});

function renderComponent(props) {
  return render(getRenderHtml(props));
}

function getRenderHtml(props) {
  return (
    <PabloThemeProvider>
      <SidebarNav onSelect={onSelectMock} {...props}>
        <SidebarNavItem>Hello</SidebarNavItem>
        <SidebarNavItem selected>World</SidebarNavItem>
        <SidebarNavItem>Bye</SidebarNavItem>
      </SidebarNav>
    </PabloThemeProvider>
  );
}
