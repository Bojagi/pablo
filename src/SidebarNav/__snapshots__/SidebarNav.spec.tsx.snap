// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render multiple nav items 1`] = `
.c0 {
  padding-left: 16px;
  border-left: 1px solid #00000020;
}

.c2 {
  font-family: "IBM Plex Sans",sans-serif;
  font-style: normal;
  font-weight: normal;
  margin: 0;
  font-size: 0.875rem;
  line-height: 1.29em;
  margin-bottom: 0;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 1px solid transparent;
  background: transparent;
  border-radius: 4px;
  -webkit-transition: background-color,0.3s,border-color 0.3s,box-shadow 0.3s ease-in-out;
  transition: background-color,0.3s,border-color 0.3s,box-shadow 0.3s ease-in-out;
  outline: none;
  margin: 8px 0;
}

.c1:not(:disabled) {
  cursor: pointer;
}

.c1:disabled {
  opacity: 0.4;
  cursor: normal;
}

.c1:hover {
  background-color: #EBE7FF;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 1px solid transparent;
  background: transparent;
  border-radius: 4px;
  -webkit-transition: background-color,0.3s,border-color 0.3s,box-shadow 0.3s ease-in-out;
  transition: background-color,0.3s,border-color 0.3s,box-shadow 0.3s ease-in-out;
  outline: none;
  margin: 8px 0;
  background-color: #EBE7FF;
}

.c3:not(:disabled) {
  cursor: pointer;
}

.c3:disabled {
  opacity: 0.4;
  cursor: normal;
}

.c3:hover {
  background-color: #EBE7FF;
}

<div>
  <ul
    class="c0"
    data-testid="pbl-sidebarnav"
  >
    <li
      class="c1"
      data-testid="pbl-sidebarnav-item"
    >
      <span
        class="c2"
      >
        Hello
      </span>
    </li>
    <li
      class="c3"
      data-testid="pbl-sidebarnav-item"
    >
      <span
        class="c2"
      >
        World
      </span>
    </li>
    <li
      class="c1"
      data-testid="pbl-sidebarnav-item"
    >
      <span
        class="c2"
      >
        Bye
      </span>
    </li>
  </ul>
</div>
`;

exports[`Render one nav item 1`] = `
.c0 {
  padding-left: 16px;
  border-left: 1px solid #00000020;
}

.c2 {
  font-family: "IBM Plex Sans",sans-serif;
  font-style: normal;
  font-weight: normal;
  margin: 0;
  font-size: 0.875rem;
  line-height: 1.29em;
  margin-bottom: 0;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 1px solid transparent;
  background: transparent;
  border-radius: 4px;
  -webkit-transition: background-color,0.3s,border-color 0.3s,box-shadow 0.3s ease-in-out;
  transition: background-color,0.3s,border-color 0.3s,box-shadow 0.3s ease-in-out;
  outline: none;
  margin: 8px 0;
  background-color: #EBE7FF;
}

.c1:not(:disabled) {
  cursor: pointer;
}

.c1:disabled {
  opacity: 0.4;
  cursor: normal;
}

.c1:hover {
  background-color: #EBE7FF;
}

<div>
  <ul
    class="c0"
    data-testid="pbl-sidebarnav"
  >
    <li
      class="c1"
      data-testid="pbl-sidebarnav-item"
    >
      <span
        class="c2"
      >
        Hello
      </span>
    </li>
  </ul>
</div>
`;
